src/docs/binding-generics.md

```diff
@@ -437,3 +437,191 @@ We could also consider solving this on the runtime level.
+        var nativeHandleSecond = Runtime.GetPayload(ref second);
+
+        var printableProtocolWitnessTableT = T.GetIPrintableProtocolWitnessTable();
+        var printableProtocolWitnessTableU = U.GetIPrintableProtocolWitnessTable();
+
+        PairPInvokes.Pair(swiftIndirectResult, nativeHandleFirst, nativeHandleSecond, firstMetadata, secondMetadata, printableProtocolWitnessTableT, printableProtocolWitnessTableU);
+    }
+
+    public static TypeMetadata Metadata
+    {
+        /* This should be cached */
+        get
+        {
+            var firstMetadata = GetTypeMetadataOrThrow<T>();
+            var secondMetadata = GetTypeMetadataOrThrow<U>();
+
+            var printableProtocolWitnessTableT = T.GetIPrintableProtocolWitnessTable();
+            var printableProtocolWitnessTableU = U.GetIPrintableProtocolWitnessTable();
+
+            // Might be handled differently due to the metadata lowering https://github.com/dotnet/runtimelab/pull/2810
+
+            return PairPInvokes.PInvokeMetadata(TypeMetadataRequest.Complete, firstMetadata, secondMetadata, printableProtocolWitnessTableT, printableProtocolWitnessTableU);
+        }
+    }
+
+    // Other members
+}
+```
+
+### Supporting generic constraints on types present in both C\# and Swift
+
+When projecting a Swift type into C\#, we can project all of the protocols it conforms to as interfaces and make the C\# equivalent implement those interfaces. This allows it to work with the generic constraints mechanism described above.
+
+However, types that exist in both languages require special handling—for example, primitive number types. In Swift, primitive number types conform to multiple protocols that their C\# counterparts do not.
+
+Consider the following Swift code:
+
+```swift
+public func acceptHashable<T: Hashable>(T data) {}
+acceptHashable(data: 3)
+```
+
+This would map into c# as described above into:
+
+```csharp
+public interface IHashable
+{
+    public abstract static NativeHandle GetIHashableProtocolWitnessTable();
+}
+
+public static void AcceptHashable<T>(T data) where T : ISwiftObject, IHashable
+{
+    var metadata = TypeMetadata.GetTypeMetadataOrThrow<T>();
+    var protocolWitnessTable = T.GetIHashableProtocolWitnessTable();
+    PInvokeAcceptHashable(data.GetPayload(), metadata, protocolWitnessTable);
+}
+
+[DllImport(Path, EntryPoint = "...")]
+[UnmanagedCallConv(CallConvs = [typeof(CallConvSwift)])]
+private static extern void PInvokeAcceptHashable(NativeHandle data, TypeMetadata metadata, NativeHandle protocolWitnessTable);
+```
+
+We would be able to call `AcceptHashable<T>` using a type that implements both `ISwiftObject` and `IHashable`, but we would not be able to call it using, for example, `System.Int64`.
+
+One way to solve this problem is by creating a wrapper struct for each of the types, which would implement the necessary interfaces:
+
+```csharp
+public struct SwiftIntWrapper : ISwiftObject, IHashable, ...
```

Comment:
I assume this would return Swift metadata for `Int`, right? So technically we could have multiple C# types which all look like `Int` from Swift's point of view (there's nothing bad about it... just trying to understand).

Reply:
Yes, you are getting this right.  As long as we abide to the calling convention and pass the metadata for Swift's `Int`, Swift should treat it as an int.

Reply:
Technically this makes another assumption - that the binary representation of value of a frozen struct with a single `Int` field is the same as of `Int` itself. Again, nothing wrong about it, but we should probably note it somewhere to make it explicit.

Reply:
>Technically this makes another assumption - that the binary representation of value of a frozen struct with a single Int field is the same as of Int itself.

Yes, thats true. Will add this assumption to the doc.

Reply:
> and we already baked in that assumption in the runtime.

Agreed - I don't think it's a problem, a comment would just help with readability.

On a related note: What if we need to do this for a ref type? Something which is a first class citizen in C# BCL, so we can't add interfaces to it, but it's a ref type. (Can't think of anything right now, so maybe it's not a problem).